tools/database/src/utils.rs

```diff
@@ -49,3 +65,190 @@ pub(crate) fn resolve_column(col_name: &str) -> anyhow::Result<DBCol> {
         .next()
         .ok_or_else(|| anyhow!("column {col_name} does not exist"))
 }
+
+pub struct MemtrieStateTrimmingCalculationResult {
+    /// State column entries that we need to keep in order to load memtries.
+    pub state_entries: Vec<(Vec<u8>, Vec<u8>)>,
+    /// A reasonable partitioning of the FlatState column keys. It is guaranteed
+    /// that these are non-overlapping, and cover the entire range of keys corresponding
+    /// to all the current shards.
+    ///
+    /// This is used to perform a parallel copy of FlatState to a fresh database.
+    pub flat_state_ranges: Vec<(Vec<u8>, Vec<u8>)>,
+}
+
+/// Reads all data from the State column that is necessary to load memtrie based on flat storage,
+/// for all shards of the current sharding version. These include the values that are not inlined,
+/// as well as some top trie nodes that are used during parallel memtrie loading. See the top of
+/// parallel_loader module for a description of the parallel loading algorithm and why these nodes
+/// are needed.
+///
+/// Also return a reasonable partitioning of the FlatState column keys.
+///
+/// Despite returning Result, for readability and ease of implementation, this function will panic
+/// for some errors that are not meaningful to recover from anyway.
+pub fn prepare_memtrie_state_trimming(
+    store: Store,
+    genesis_config: &GenesisConfig,
+    include_flat_delta: bool,
+) -> anyhow::Result<MemtrieStateTrimmingCalculationResult> {
+    let epoch_manager = EpochManager::new_arc_handle(store.clone(), genesis_config);
+    let chain = ChainStore::new(store.clone(), genesis_config.genesis_height, false);
+    let final_head = chain.final_head()?;
+    let epoch_id = final_head.epoch_id;
+    let shard_layout = epoch_manager.get_shard_layout(&epoch_id)?;
+    let shard_uids = shard_layout.shard_uids().collect::<Vec<_>>();
+
+    let mut flat_state_ranges = Vec::new();
+
+    // Spawn a thread to collect all the important keys that we'll discover.
+    let (important_keys_tx, important_keys_rx) = std::sync::mpsc::sync_channel::<Vec<u8>>(1000000);
+    let important_keys_collect_thread = std::thread::spawn(move || {
+        let mut important_keys = Vec::new();
+        while let Ok(key) = important_keys_rx.recv() {
+            important_keys.push(key);
+        }
+        important_keys
+    });
+    for shard_uid in shard_uids {
```

Comment:
not a big deal to do this now, but do we also want to parallelize this part? we might wait for a while working on a shard that doesnt fully utilize all cores, when we could be making progress on other shards too

Reply:
It wouldn't be of much help. This part is negligible compared to the copying-over-of-flat-state part. In practice it'll take at most a few seconds each, whereas the copying will take a few minutes.